community-containers/stalwart/stalwart.json

```diff
@@ -38,6 +38,11 @@
                   "ip_binding": "",
                   "port_number": "4190",
                   "protocol": "tcp"
+                },
+                {
+                  "ip_binding": "",
+                  "port_number": "8080",
```

Comment:
I fear this will conflict with AIO's port. In best case the port that gets exposed by Stalwart should be changed to something else. However not sure if possible.

Reply:
What is the syntax for this?
And, should put a binding IP to "%APACHE_IP_BINDING%" ?

Reply:
I dont know syntax too, so I did my experiments with standalone containers.
Maybe Simon help.

Reply:
> Stalwart-mail docker port can be mapped from 8080 to 10003 for example, similiar as apache container do.
Caddy can proxy mail.$ncdomain - > localhost:10003

Sounds good! How can we adjust the port inside stalwart? Is ther an environmental vairable for this?

Reply:
> What is the syntax for this?
> And, should put a binding IP to "%APACHE_IP_BINDING%" ?

For this port indeed I would recommend to use `%APACHE_IP_BINDING%`.

Reply:
I've made some changes and now have a Stalwart server up and running.
In the mastercontainer, in the stalwart.json file, add 2 ports. You can do it with:
sudo docker exec -it nextcloud-aio-mastercontainer vi /var/www/docker-aio/community-containers/stalwart/stalwart.json
{
"ip_binding": "",
"port_number": "8082",
"protocol": "tcp"
},
{
"ip_binding": "",
"port_number": "4433",
"protocol": "tcp"
}
And inside the Stalwart container, modify the config.toml file. You can do it with:
sudo docker exec -it nextcloud-aio-stalwart vi /opt/stalwart-mail/etc/config.toml
And replace the line
server.listener.http.bind.0 = "[::]:8080"
with
server.listener.http.bind.0 = "[::]:8082"
And the line
server.listener.https.bind = "[::]:443"
with 
server.listener.https.bind = "[::]:4433"

You need to restart the master and stalwart containers and open your firewall/router ports

I think all those changes could be done by modifying community-containers/stalwart/stalwart.json in this project and the start.sh file in the project https://github.com/marcoambrosini/aio-stalwart/blob/main/start.sh

Reply:
Cannt mastercontainer spawn stalwart-mail as `docker run -d -p 8082:8080` and let caddy to serve it?

Reply:
What configuration?

Reply:
If the certificates are always renewed on the same path then no need to update Stalwart.
> `cert`: The path or content of the TLS certificate file. This can either be directly embedded in the configuration file or referenced from an external file using a [file macro](https://stalw.art/docs/configuration/macros).
> `private-key`: The path or content of the TLS private key file. This should not be embedded directly in the configuration file, and instead be referenced from an external file using a [file macro](https://stalw.art/docs/configuration/macros).
https://stalw.art/docs/server/tls/certificates

Reply:
![image](https://github.com/nextcloud/all-in-one/assets/7360784/b43834aa-08b7-4485-8062-3e92916af559)
it looks like this. Easy to do.
Advanced user can use their own cert

Reply:
The anothe way to define certificates is config.toml like this.

```
cat >> /opt/stalwart-mail/etc/config.toml <<EOF
server.tls.certificate = "default"
cert = "file:///caddy/caddy/certificates/acme-v02.api.letsencrypt.org-directory/mail.$NC_DOMAIN/mail.$NC_DOMAIN.crt"
private-key = "file:///caddy/caddy/certificates/acme-v02.api.letsencrypt.org-directory/mail.$NC_DOMAIN/mail.$NC_DOMAIN.key"
certificate.default.default = true
EOF

```

Reply:
and `sed -i "s/[::]:8080/[::]:8082/g"`
with json
```json
{
"ip_binding": "%APACHE_IP_BINDING%",
"port_number": "8082",
"protocol": "tcp"
}
```
I'm not sure if it should be http or https.

Reply:
> sed -i "s/[::]:8080/[::]:8082/g"

Can you send this as a PR to https://github.com/marcoambrosini/aio-stalwart?

Reply:
I don't think they'll want to change the default port. It's configurable, so it's up to us to change the configuration...

Reply:
This container is customized for aio so it will be fine

Reply:
Sorry, I understood on the Stalwart repository
I made a PR : https://github.com/marcoambrosini/aio-stalwart/pull/13#issue-2269263516